src/vos/vos_internal.h

```diff
@@ -365,64 +365,79 @@ extern uint64_t vos_evt_feats;
+/** Left shift for stored portion of epoch */
+#define VOS_AGG_LSHIFT	(VOS_AGG_NR_BITS - VOS_AGG_RSHIFT)
 /** In-place mask to get epoch from feature bits */
-#define VOS_TF_AGG_TIME_MASK	(0xffffffffULL << VOS_TF_AGG_LSHIFT)
+#define VOS_AGG_TIME_MASK	(VOS_AGG_EPOCH_MASK >> VOS_AGG_RSHIFT)
 
-D_CASSERT((VOS_TF_AGG_HLC & VOS_TF_AGG_TIME_MASK) == 0);
-D_CASSERT(VOS_TF_AGG_TIME_MASK & (1ULL << VOS_TF_AGG_BIT));
-D_CASSERT((VOS_TF_AGG_TIME_MASK & (1ULL << (VOS_TF_AGG_BIT + 1))) == 0);
-D_CASSERT((VOS_TF_AGG_TIME_MASK & (1ULL << (VOS_TF_AGG_BIT - 32))) == 0);
+D_CASSERT((VOS_TF_AGG_HLC & VOS_AGG_TIME_MASK) == 0);
+D_CASSERT(VOS_AGG_TIME_MASK & (1ULL << VOS_TF_AGG_BIT));
+D_CASSERT((VOS_AGG_TIME_MASK & (1ULL << (VOS_TF_AGG_BIT + 1))) == 0);
+D_CASSERT((VOS_AGG_TIME_MASK & (1ULL << (VOS_TF_AGG_BIT - VOS_AGG_NR_BITS))) == 0);
 
 #define CHECK_VOS_TREE_FLAG(flag)	\
 	D_CASSERT(((flag) & (EVT_FEATS_SUPPORTED | BTR_FEAT_MASK)) == 0)
 CHECK_VOS_TREE_FLAG(VOS_KEY_CMP_LEXICAL);
 CHECK_VOS_TREE_FLAG(VOS_TF_AGG_OPT);
-CHECK_VOS_TREE_FLAG(VOS_TF_AGG_TIME_MASK);
+CHECK_VOS_TREE_FLAG(VOS_AGG_TIME_MASK);
 
-/** Update the aggregatable write timestamp within 1/4 second granularity */
-static inline void
-vos_feats_agg_time_update(daos_epoch_t epoch, uint64_t *feats)
+/** Get the aggregatable write timestamp within 1/4 ms granularity */
+static inline bool
+vos_feats_agg_time_get(uint64_t feats, daos_epoch_t *epoch)
 {
-	uint64_t extra_flag = 0;
+	if ((feats & VOS_TF_AGG_OPT) == 0)
+		return false;
 
-	if (epoch & 0xffffffff00000000ULL) {
-		/** Assume aggregation only happens at most once every quarter second */
-		epoch &= 0xffffffff00000000ULL;
-		/** Ensure the resulting epoch is not zero regardless, mostly for standalone */
-		epoch = (epoch >> VOS_TF_AGG_RSHIFT);
-		extra_flag = VOS_TF_AGG_HLC;
-	} else {
-		epoch &= 0x0ffffffffULL;
-		epoch = (epoch << VOS_TF_AGG_LSHIFT);
-	}
+	if (feats & VOS_TF_AGG_HLC)
+		*epoch = ((feats & VOS_AGG_TIME_MASK) << VOS_AGG_RSHIFT);
+	else
+		*epoch = ((feats & VOS_AGG_TIME_MASK) >> VOS_AGG_LSHIFT);
 
-	*feats &= ~VOS_TF_AGG_TIME_MASK;
-	*feats |= epoch | VOS_TF_AGG_OPT | extra_flag;
+	return true;
 }
 
-/** Get the aggregatable write timestamp within 1/4 second granularity */
-static inline bool
-vos_feats_agg_time_get(uint64_t feats, daos_epoch_t *epoch)
+/** Update the aggregatable write timestamp within 1/4 ms granularity */
+static inline void
+vos_feats_agg_time_update(daos_epoch_t epoch, uint64_t *feats)
 {
-	if ((feats & VOS_TF_AGG_OPT) == 0)
-		return false;
+	uint64_t	extra_flag = 0;
+	daos_epoch_t	old_epoch = 0;
 
-	if (feats & VOS_TF_AGG_HLC) {
-		*epoch = ((feats & VOS_TF_AGG_TIME_MASK) << VOS_TF_AGG_RSHIFT);
-		*epoch += 0xffffffffULL;
+	if (!vos_feats_agg_time_get(*feats, &old_epoch))
+		old_epoch = 0;
+
+	if (epoch <= old_epoch) /** Only need to save newest */
+		return;
+
+	if (epoch & VOS_AGG_HLC_MASK) {
+		/** We only save the top bits so round up so the stored timestamp works */
+		epoch += VOS_AGG_HLC_BITS;
```

Comment:
Why do we need to add the lower bits on updating? Shouldn't it being added on getting?

Reply:
I'm just rounding up to nearest 1/4 ms.    If I had done it this way to start with, it would be forward compatible.

Reply:
Ok, I see. Thanks.